Add auth spans and remove deduplication code for scopes

jalevin opened the pull request:
**What is this feature?**
This PR adds more spans for timing in accesscontrol. We used the timing insights to investigate some performance bottlenecks in access control phase of requests and found the deduplication of permissions was the culprit. After consulting with @grafana/identity-access-team we decided to remove this.

jalevin committed:
pass slice for allocations

jalevin committed:
test if revert fixes e2e tests

jalevin committed:
refactor

hairyhenderson committed:
chore: remove unrelated formatting

Signed-off-by: Dave Henderson <dave.henderson@grafana.com>

jalevin committed:
Update pkg/services/accesscontrol/acimpl/service.go

Co-authored-by: Dave Henderson <dave.henderson@grafana.com>

jalevin committed:
fix span relationship in auth and add span to getDashboard

jalevin committed:
remove comment

jalevin committed:
fix format

jalevin committed:
fix naming and formatting tweaks

jalevin committed:
add more spans and attributes on permissions / timing

jalevin committed:
revert formatting

jalevin committed:
add tracer to http server for dashboard tests

jalevin committed:
formatting

jalevin committed:
closure

jalevin committed:
bump

jalevin committed:
add tests

jalevin committed:
fix missing param in test

jalevin committed:
remove closure

IevaVasiljeva commented on pkg/services/accesscontrol/acimpl/service.go:
```diff
@@ -437,6 +459,9 @@ func GetActionFilter(options accesscontrol.SearchOptions) func(action string) bo
 
 // SearchUsersPermissions returns all users' permissions filtered by action prefixes
 func (s *Service) SearchUsersPermissions(ctx context.Context, usr identity.Requester, options accesscontrol.SearchOptions) (map[int64][]accesscontrol.Permission, error) {
+	ctx, span := s.tracer.Start(ctx, "authz.SearchUserPermissions")
```
```suggestion
	ctx, span := s.tracer.Start(ctx, "authz.SearchUsersPermissions")
```

IevaVasiljeva reviewed (approved):
Oh, interesting. Deduplicating permissions is actually something that I added relatively recently (in [this PR](https://github.com/grafana/grafana/pull/87967/files#diff-3f46355ace70434e1c19bdc6f340aa299f8324a0aa87870a3aa1384c6c919d92R235)). My thinking was that deduplicating the scopes once should make any future permission lookups from the cached permissions faster. But I didn't test this change at scale, so if it's too CPU heavy we can undo it. It would also be interesting to see if deduplication does actually lead to any performance benefits - no point doing it if it doesn't.

hairyhenderson commented:
> Deduplicating permissions is actually something that I added relatively recently

There's a bunch of things going on in that PR and I haven't read very closely so apologies if it's obvious, but: is the `GroupScopesByAction` change such that it can be backed out with little impact?

Also, I'm not convinced that the deduplication _itself_ is to blame for the poor performance in the investigations that @jalevin and I have been doing, but rather the allocations happening in that loop. It's possible that we might not be able to avoid it without some fairly big changes 🤔

jalevin commented:
> Oh, interesting. Deduplicating permissions is actually something that I added relatively recently (in [this PR](https://github.com/grafana/grafana/pull/87967/files#diff-3f46355ace70434e1c19bdc6f340aa299f8324a0aa87870a3aa1384c6c919d92R235)). My thinking was that deduplicating the scopes once should make any future permission lookups from the cached permissions faster. But I didn't test this change at scale, so if it's too CPU heavy we can undo it. It would also be interesting to see if deduplication does actually lead to any performance benefits - no point doing it if it doesn't.

We added some span attributes to look at reductions. In a configuration with default permissions, 20k users, and 1m dashboards we generate 5095220 permissions. After deduplication we get 5095155 which removed 65 duplicates over 115 actions.

I ran some benchmarks with and without deduplication:

<img width="654" alt="image" src="https://github.com/grafana/grafana/assets/1669620/03f8aba2-009e-4e48-8af7-0d225709ee78">

I think part of the problem is that we're not actually caching the result of the deduplication so this is happening on every request AFTER gathering permissions from the cached sources. 

Timing of web requests:

deduplicated:
<img width="1293" alt="image" src="https://github.com/grafana/grafana/assets/1669620/65f33a27-7f66-4441-9db7-5bb79d429ebf">

non-deduplicated:
<img width="1289" alt="image" src="https://github.com/grafana/grafana/assets/1669620/90ae3cbb-bf88-41e6-844e-a34f08a958a1">

I'm not entirely sure what configuration could lead to duplicate permissions, however, I'd recommend we either:

1. disable it on permission counts > 100k
2. cache the end result of deduplicated and pull from that cache
3. remove altogether

jalevin committed:
checkpoint

jalevin committed:
remove deduplication comment code

jalevin committed:
dont use deprecated method

jalevin committed:
undo formatting

jalevin committed:
Update pkg/services/accesscontrol/acimpl/service.go

Co-authored-by: Ieva <ieva.vasiljeva@grafana.com>

jalevin commented:
Consensus reached on option 3, remove deduplication code.

jalevin committed:
fix linting errors

jalevin committed:
fix linting

jalevin committed:
typo

jalevin committed:
remove duplicate check

jalevin committed:
bump

jalevin committed:
fix test mocks

jalevin committed:
Merge branch 'main' into jalevin/add_more_auth_spans

jalevin committed:
fix

jalevin committed:
bump
